editor/gui/touch_actions_panel.cpp

```diff
@@ -0,0 +1,183 @@
+
+void TouchActionsPanel::_simulate_key_press(Key keycode) {
+	Ref<InputEventKey> event;
+	event.instantiate();
+	event->set_keycode(keycode);
+	event->set_pressed(true);
+	Input::get_singleton()->parse_input_event(event);
+}
+
+Button *TouchActionsPanel::_add_new_action_button(const String &shortcut, Key keycode) {
+	Button *action_button = memnew(Button);
+	action_button->set_focus_mode(Control::FOCUS_NONE);
+	action_button->set_h_size_flags(Control::SIZE_SHRINK_CENTER);
+	action_button->set_v_size_flags(Control::SIZE_SHRINK_CENTER);
+	if (keycode == Key::NONE) {
+		action_button->connect(SceneStringName(pressed), callable_mp(this, &TouchActionsPanel::_simulate_editor_shortcut).bind(shortcut));
+	} else {
+		action_button->connect(SceneStringName(pressed), callable_mp(this, &TouchActionsPanel::_simulate_key_press).bind(keycode));
+	}
+	box->add_child(action_button);
+	return action_button;
+}
+
+void TouchActionsPanel::_on_drag_handle_gui_input(const Ref<InputEvent> &event) {
+	if (lock_panel_position) {
+		return;
+	}
+	Ref<InputEventMouseButton> mouse_button_event = event;
+	if (mouse_button_event.is_valid() && mouse_button_event->get_button_index() == MouseButton::LEFT) {
+		if (mouse_button_event->is_pressed()) {
+			dragging = true;
+			drag_offset = mouse_button_event->get_position();
+		} else {
+			dragging = false;
+		}
+	}
+
+	Ref<InputEventMouseMotion> mouse_motion_event = event;
+	if (mouse_motion_event.is_valid() && dragging) {
+		Vector2 new_position = get_position() + mouse_motion_event->get_relative();
+		const float margin = 25.0;
+		Vector2 parent_size = get_parent_area_size();
+		Vector2 panel_size = get_size();
+		new_position.x = CLAMP(new_position.x, margin, parent_size.x - panel_size.x - margin);
+		new_position.y = CLAMP(new_position.y, margin, parent_size.y - panel_size.y - margin);
+		set_position(new_position);
+	}
+}
+
+void TouchActionsPanel::_switch_layout() {
+	box->set_vertical(!box->is_vertical());
+	if (box->is_vertical()) {
+		v_separator->hide();
+		h_separator->show();
+	} else {
+		v_separator->show();
+		h_separator->hide();
+	}
+	reset_size();
+}
+
+void TouchActionsPanel::_lock_panel_toggled(bool p_pressed) {
+	lock_panel_position = p_pressed;
+	layout_toggle_button->set_disabled(p_pressed);
+}
+
+TouchActionsPanel::TouchActionsPanel() {
+	dragging = false;
+
+	Ref<StyleBoxFlat> panel_style = memnew(StyleBoxFlat);
+	panel_style->set_bg_color(Color(0.1, 0.1, 0.1, 0.95));
+	panel_style->set_border_color(Color(0.3, 0.3, 0.3, 1));
+	panel_style->set_border_width_all(3);
+	panel_style->set_corner_radius_all(10);
+	panel_style->set_content_margin_all(12);
+	add_theme_style_override("panel", panel_style);
+
+	set_anchors_and_offsets_preset(Control::PRESET_CENTER_BOTTOM, Control::PRESET_MODE_MINSIZE, 80);
+
+	box = memnew(BoxContainer);
+	box->set_alignment(BoxContainer::ALIGNMENT_CENTER);
+	box->add_theme_constant_override("separation", 15);
+	box->set_vertical(false);
+	add_child(box);
+
+	// Add drag handle.
```

Comment:
Comments like that are not really necessary. It can be inferred from variable name already.